---
title: "Quality Control"
author: "Sergey Naumenko"
date: "`r Sys.Date()`"
output:
  pdf_document:
    toc: yes
  html_document:
    code_folding: hide
    df_print: paged
    highlights: pygments
    number_sections: yes
    self_contained: yes
    theme: default
    toc: yes
    toc_float:
      collapsed: yes
      smooth_scroll: yes
---

# Overview
- Principal Investigator: Beth Overmoyer
- Experiment: RNAseq_analysis_of_inflammatory_breast_cancer_hbc04141
- study 6 was excluded

```{r, cache = FALSE, message = FALSE, include = FALSE}
library(tidyverse)
library(knitr)
library(DESeq2)
library(DEGreport)
library(ggrepel)
library(pROC)
#devtools::install_github("sachsmc/plotROC")
#library(plotROC)

ggplot2::theme_set(theme_light(base_size = 14))

opts_chunk[["set"]](
    cache = FALSE,
    dev = c("png", "pdf"),
    error = TRUE,
    highlight = TRUE,
    message = FALSE,
    prompt = FALSE,
    tidy = FALSE,
    warning = FALSE)
```

# Metadata
```{r  show-metadata, rows.print = 46}
se <- readRDS("data/bcbio-se_all_samples.rds")
se <- se[, se$study_id != 6]
mdata <- colData(se) 
mdata <- subset(mdata, subset = study_id != 6)
mdata$response <- str_replace(mdata$response, "Yes", "pCR")
mdata$response <- str_replace(mdata$response, "No", "non-pCR")
mdata$response <- replace_na(mdata$response, "non-pCR")
colData(se) <- mdata

mdata <- metadata(se)
mdata$metrics <- mdata$metrics %>% dplyr::filter(!sample %in% c("s3370_1", "s3373_3"))
metadata(se) <- mdata

metadata <- colData(se) %>% 
    as_tibble(rownames = NULL) %>% 
    dplyr::select(-batch, -phenotype)

metrics <- metadata(se)$metrics %>% 
    left_join(metadata, by = c("sample" = "sample"))

metrics$date_of <- as_factor(metrics$date_of)

metadata
```

\newpage

# Read metrics {.tabset}

## Total reads
```{r plot_total_reads, fig.width = 10, fig.height = 20}
metrics <- metrics %>% 
    group_by(date_of) %>% 
    mutate(position = rank(-total_reads))

metrics %>%
    ggplot(aes(x = reorder(sample, -total_reads),
               y = total_reads/1e6L, 
               fill = date_of,
               group = position)) +
    geom_bar(stat = "identity") +
    coord_flip() +
    scale_y_continuous(name = "mln reads") +
    geom_text(aes(label = floor(total_reads/1e6L)), hjust = 0, nudge_y = 0.5)+
    xlab("sample") + 
    ggtitle("Total reads")
```

## Mapped reads
The number of mapped reads should correspond to the number of total reads.
```{r plot_mapped_reads, fig.width = 10, fig.height = 20}
metrics %>%
    ggplot(aes(x = reorder(sample, -mapped_reads),
               y = mapped_reads, fill = date_of)) +
    geom_bar(stat = "identity") +
    coord_flip() +
    scale_y_continuous(name = "mln reads") +
    geom_text(aes(label = floor(mapped_reads/1e6L)), hjust = 0, nudge_y = 0.5)+
    xlab("sample") + 
    ggtitle("Mapped reads")
```

```{r}
## Mapping rate
#The genomic mapping rate represents the percentage of reads mapping to the reference genome. Low mapping rates are indicative of sample #contamination, poor sequencing quality or other artifacts.
#{r plot_mapping_rate, fig.width = 10, fig.height = 20}
#metrics %>%
#    ggplot(aes(x = reorder(sample, -mapped_reads_pct), 
#               y = mapped_reads_pct, fill = tissue)) +
#        geom_bar(stat = "identity") +
#    coord_flip() +
#    geom_text(aes(label = floor(mapped_reads_pct)), hjust = 0, nudge_y = 0.5)+
#    xlab("sample") + 
#    ggtitle("Mapping rate")
```


## Number of genes detected
```{r plot_genes_detected, fig.width = 10, fig.height = 20}
genes_detected <- colSums(assays(se)[["raw"]] > 0) %>% enframe()
colnames(genes_detected) <- c("sample", "n_genes")

metrics <- metrics %>%
    left_join(genes_detected, by = c("sample" = "sample"))

metrics %>% 
    ggplot(aes(x = reorder(sample, -n_genes),
               y = n_genes, fill = date_of)) +
    geom_bar(stat = "identity") +
    coord_flip() +
    geom_text(aes(label = n_genes), hjust = 0, nudge_y = 0.5)+
    xlab("sample") + 
    ylim(0, 50000) +
    ggtitle("Number of genes")
```


## Gene detection saturation

We should observe a linear trend in the number of genes detected with the number of mapped reads, which indicates that the sample input was not overloaded.

```{r plot_gene_saturation, fig.width = 20, fig.height = 20}
metrics %>% 
    ggplot(aes(x = log10(total_reads), 
               y = n_genes,
               color = date_of)) +
        geom_point(size = 5)+
        geom_text_repel(aes(label = sample), size = 10)+
    theme(axis.text = element_text(size = 20),
          axis.title = element_text(size = 20, face="bold"),
          legend.text = element_text(size = rel(1))) +
    theme_bw(base_size = 20) +
    ggtitle("Gene saturation")
```

## Exonic mapping rate
Ideally, at least 60% of total reads should map to exons.
```{r plot_exonic_mapping_rate, fig.width = 10, fig.height = 20}
metrics %>%
    ggplot(aes(x = reorder(sample, -exonic_rate),
               y = exonic_rate * 100, 
               fill = date_of)) +
        geom_bar(stat = "identity") +
    geom_text(aes(label = floor(exonic_rate*100)), hjust = 0, nudge_y = 0.5)+
    xlab("sample") + 
    ylab("Exonic rate %") + 
    ggtitle("Exonic mapping rate") + 
    coord_flip() 
```

## Intronic mapping rate
The majority of reads should map to exons and not introns.
```{r plot_intronic_mapping_rate, fig.width = 10, fig.height = 20}
metrics %>%
    ggplot(aes(x = reorder(sample, -intronic_rate),
               y = intronic_rate * 100, 
               fill = date_of)) +
        geom_bar(stat = "identity") +
        coord_flip() +
        geom_text(aes(label = floor(intronic_rate*100)), 
                  hjust = 0, 
                  nudge_y = 0.5) +
        xlab("sample") + 
        ylab("Intronic rate %")
        ggtitle("Intronic mapping rate")
```

## rRNA mapping rate
Samples should have a ribosomal RNA (rRNA) contamination rate below 10%.

```{r plot_rrna_mapping_rate, fig.width = 10, fig.height = 20}
metrics %>%
    ggplot(aes(x = reorder(sample, -r_rna_rate),
               y = r_rna_rate * 100, 
               fill = date_of)) +
        geom_bar(stat = "identity") +
    geom_text(aes(label = round(r_rna_rate*100,2)), hjust = 0, nudge_y = 0.5)+
    xlab("sample") + 
    ylab("rRNA rate, %")+
    ylim(0, 1) + 
    ggtitle("rRNA mapping rate") + 
    coord_flip() 
```

## 5'->3' bias
```{r plot_53_bias, fig.width = 10, fig.height = 20}
metrics %>%
    ggplot(aes(x = reorder(sample, -x5_3_bias),
               y = x5_3_bias, 
               fill = date_of)) +
        geom_bar(stat = "identity") +
    geom_text(aes(label = x5_3_bias), hjust = 0, nudge_y = 0.01)+
    xlab("sample") + 
    ylim(0, 1.5)+
    ggtitle("5'-3' bias") + 
    coord_flip()

```

## Counts per gene - all genes
We expected similar spread for every sample.
```{r plot_counts_per_gene, fig.width = 10, fig.height = 20}
metadata$date_of <- as_factor(metadata$date_of)
metrics_small <- metrics %>% dplyr::select(sample, date_of)
assays(se)[["raw"]] %>% 
    as_tibble() %>% 
    dplyr::filter(rowSums(.)!=0) %>% 
    gather(sample, counts) %>% 
    left_join(metadata, by = c("sample" = "sample")) %>% 
    ggplot(aes(sample, log2(counts+1), fill = date_of)) +
    geom_boxplot() + 
    coord_flip() +
    ggtitle("Counts per gene, all non-zero genes")
```

## Counts per gene - protein coding genes
```{r plot_counts_per_gene_protein_coding, fig.width = 10, fig.height = 20}
protein_coding_genes <- read_csv("tables/ensembl_w_description.protein_coding.csv")
assays(se)[["raw"]] %>% 
    as_tibble(rownames = "ensembl_gene_id") %>% 
    dplyr::filter(ensembl_gene_id %in% protein_coding_genes$ensembl_gene_id) %>% 
    dplyr::select(-ensembl_gene_id) %>% 
    dplyr::filter(rowSums(.)!=0) %>% 
    gather(sample, counts) %>% 
    left_join(metadata, by = c("sample" = "sample")) %>% 
    dplyr::mutate(date_of = as.factor(date_of)) %>% 
    ggplot(aes(sample, log2(counts+1), fill = date_of)) +
    geom_boxplot() + 
    coord_flip() +
    ggtitle("Counts per gene, protein_coding_genes")
```

# Sample similarity analysis

## Principal component analysis (PCA) - non zero genes 

```{r pca, fig.width = 20, fig.height = 20}
raw_counts <- assays(se)[["raw"]] %>% 
    as_tibble() %>% 
    dplyr::filter(rowSums(.)!=0) %>% 
    as.matrix()

vst <- vst(raw_counts) 

pca <- degPCA(vst, colData(se), condition = "date_of", name = "sample", data = T)[["plot"]]
pca_labels <- pca[["labels"]]
pca_data <- pca[["data"]] %>% as_tibble() %>%
    dplyr::select(sample, PC1, PC2, date_of)

pca_data$date_of <- as.factor(pca_data$date_of)

pca_data %>% 
    ggplot(aes(x = PC1, y = PC2, color = date_of, label = sample)) + 
    geom_point(size = 5) + 
    geom_text_repel(size = 10) + 
    xlab(pca_labels$x) + 
    ylab(pca_labels$y) +
    ggtitle(paste0("PCA: ", nrow(vst), " genes")) +
    theme_bw(base_size = 20) +
    theme(axis.text = element_text(size = 20),
          axis.title = element_text(size = 20, face="bold"),
          legend.text = element_text(size = rel(1)))
```

## Principal component analysis (PCA) - protein coding genes

```{r pca_coding_filtered, fig.width = 20, fig.height = 20}
raw_counts <- assays(se)[["raw"]] %>% 
    as_tibble(rownames = "ensembl_gene_id") %>% 
    dplyr::filter(ensembl_gene_id %in% protein_coding_genes$ensembl_gene_id) %>% 
    column_to_rownames(var = "ensembl_gene_id") %>% 
    dplyr::filter(rowSums(.)!=0) %>% 
    as.matrix()

vst <- vst(raw_counts) 

pca <- degPCA(vst, colData(se), condition = "date_of", name = "sample", data = T)[["plot"]]
pca_labels <- pca[["labels"]]
pca_data <- pca[["data"]] %>% as_tibble() %>%
    dplyr::select(sample, PC1, PC2, date_of)

pca_data$date_of <- as.factor(pca_data$date_of)

pca_data %>%
    ggplot(aes(x = PC1, y = PC2, color = date_of, label = sample)) + 
    geom_point(size = 5) + 
    geom_text_repel(size = 10) + 
    xlab(pca_labels$x) + 
    ylab(pca_labels$y) +
    ggtitle(paste0("PCA:", nrow(vst), " genes")) +
    theme_bw(base_size = 20) +
    theme(axis.text = element_text(size = 20),
          axis.title = element_text(size = 20, face="bold"),
          legend.text = element_text(size = rel(1)))

```
    
    
## Principal component analysis (PCA) - protein coding genes - er

```{r pca_coding_filtered_er, fig.width = 20, fig.height = 20}
raw_counts <- assays(se)[["raw"]] %>% 
    as_tibble(rownames = "ensembl_gene_id") %>% 
    dplyr::filter(ensembl_gene_id %in% protein_coding_genes$ensembl_gene_id) %>% 
    column_to_rownames(var = "ensembl_gene_id") %>% 
    dplyr::filter(rowSums(.)!=0) %>% 
    as.matrix()

vst <- vst(raw_counts) 

pca <- degPCA(vst, colData(se), condition = "er", name = "sample", data = T)[["plot"]]
pca_labels <- pca[["labels"]]
pca_data <- pca[["data"]] %>% as_tibble() %>%
    dplyr::select(sample, PC1, PC2, er)

pca_data$er <- as.factor(pca_data$er)

pca_data %>% 
    ggplot(aes(x = PC1, y = PC2, color = er, label = sample)) + 
    geom_point(size = 5) + 
    geom_text_repel(size = 10) + 
    xlab(pca_labels$x) + 
    ylab(pca_labels$y) +
    ggtitle(paste0("PCA:", nrow(vst), " genes")) +
    theme_bw(base_size = 20) +
    theme(axis.text = element_text(size = 20),
          axis.title = element_text(size = 20, face="bold"),
          legend.text = element_text(size = rel(1)))

```


\newpage

## Principal component analysis (PCA) - protein coding genes - er

```{r pca_coding_filtered_studyid, fig.width = 20, fig.height = 20}
raw_counts <- assays(se)[["raw"]] %>% 
    as_tibble(rownames = "ensembl_gene_id") %>% 
    dplyr::filter(ensembl_gene_id %in% protein_coding_genes$ensembl_gene_id) %>% 
    column_to_rownames(var = "ensembl_gene_id") %>% 
    dplyr::filter(rowSums(.)!=0) %>% 
    as.matrix()

vst <- vst(raw_counts) 

pca <- degPCA(vst, colData(se), condition = "study_id", name = "sample", data = T)[["plot"]]
pca_labels <- pca[["labels"]]
pca_data <- pca[["data"]] %>% as_tibble() %>%
    dplyr::select(sample, PC1, PC2, study_id)

pca_data$study_id <- as.factor(pca_data$study_id)

pca_data %>% 
    ggplot(aes(x = PC1, y = PC2, color = study_id, label = study_id)) + 
    geom_point(size = 5) + 
    geom_text_repel(size = 10) + 
    xlab(pca_labels$x) + 
    ylab(pca_labels$y) +
    ggtitle(paste0("PCA:", nrow(vst), " genes")) +
    theme_bw(base_size = 20) +
    theme(axis.text = element_text(size = 20),
          axis.title = element_text(size = 20, face="bold"),
          legend.text = element_text(size = rel(1)))

```

\newpage

# PCA - protein coding genes - PRE/POST - Responce

## PRE
```{r pca_pre, fig.width = 10, fig.height = 10}
se_pre <- se[,se$treatment == "pre"]
raw_counts <- assays(se_pre)[["raw"]] %>% 
    as_tibble(rownames = "ensembl_gene_id") %>% 
    dplyr::filter(ensembl_gene_id %in% protein_coding_genes$ensembl_gene_id) %>% 
    column_to_rownames(var = "ensembl_gene_id") %>% 
    dplyr::filter(rowSums(.)!=0) %>% 
    as.matrix()

vst <- vst(raw_counts) 

pca <- degPCA(vst, colData(se_pre), condition = "response", name = "sample", data = T)[["plot"]]
pca_labels <- pca[["labels"]]
pca_data_pre <- pca[["data"]] %>% as_tibble() %>%
    dplyr::select(sample, PC1, PC2, response)

pca_data_pre$response <- as.factor(pca_data_pre$response)

# ggtitle(paste0("PCA:", nrow(vst), " genes")) +
fig3b <- pca_data_pre %>%
    ggplot(aes(x = PC1, y = PC2, color = response, label = sample)) + 
    geom_point(size = 5) + 
    geom_text_repel(size = 10) + 
    xlab(pca_labels$x) + 
    ylab(pca_labels$y) +
    theme_bw(base_size = 15) +
    theme(axis.text = element_text(size = 20),
          axis.title = element_text(size = 20, face="bold"),
          legend.text = element_text(size = rel(1)),
          panel.border = element_blank(),
          panel.grid.minor = element_blank(),
          panel.grid.major = element_blank())

saveRDS(fig3b, "data/fig3b.RDS")
fig3b
```

\newpage

## PRE - ROC
```{r}
#pca_data$PC2
#pca_data$response
roc_pre <- roc(pca_data_pre$response, pca_data_pre$PC2, percent = FALSE)
ci.auc(roc_pre)
ggroc(roc_pre, 
      colour = "red", linetype = 2, size = 1, legacy.axes = TRUE)+
    theme_minimal() + 
  scale_x_continuous(labels = scales::percent) +
  scale_y_continuous(labels = scales::percent) +
  xlab("100% - specificity : False Positive Percentage") +
  ylab("Sensitivity : True Postive Percentage") + 
  ggtitle(paste0("PRE AUC : ", round(roc_pre$auc, 2)))
```



\newpage

## POST
```{r pca_post, fig.width = 10, fig.height = 10}
se_post <- se[,se$treatment == "post"]
raw_counts <- assays(se_post)[["raw"]] %>% 
    as_tibble(rownames = "ensembl_gene_id") %>% 
    dplyr::filter(ensembl_gene_id %in% protein_coding_genes$ensembl_gene_id) %>% 
    column_to_rownames(var = "ensembl_gene_id") %>% 
    dplyr::filter(rowSums(.)!=0) %>% 
    as.matrix()

vst <- vst(raw_counts) 

pca_post <- degPCA(vst, colData(se_post), condition = "response", name = "sample", data = T)[["plot"]]
pca_labels <- pca_post[["labels"]]
pca_data_post <- pca_post[["data"]] %>% as_tibble() %>%
    dplyr::select(sample, PC1, PC2, response)

pca_data_post$response <- as.factor(pca_data_post$response)

#ggtitle(paste0("PCA:", nrow(vst), " genes")) +
fig3c <- pca_data_post %>%
    ggplot(aes(x = PC1, y = PC2, color = response, label = sample)) + 
    geom_point(size = 5) + 
    geom_text_repel(size = 10) + 
    xlab(pca_labels$x) + 
    ylab(pca_labels$y) +
    theme_bw(base_size = 15) +
    theme(axis.text = element_text(size = 20),
          axis.title = element_text(size = 20, face="bold"),
          legend.text = element_text(size = rel(1)),
          panel.border = element_blank(),
          panel.grid.minor = element_blank(),
          panel.grid.major = element_blank())

saveRDS(fig3c, "data/fig3c.RDS")
fig3c
```

\newpage

## POST - ROC
```{r, fig.width = 10, fig.height = 10}
#pca_data_post$PC2
#pca_data_post$response
# https://sachsmc.github.io/plotROC/
roc_post <- roc(pca_data_post$response, pca_data_post$PC2, percent = FALSE)
ci.auc(roc_post)

fig3d <- pROC::ggroc(list(pre = roc_pre,
           post = roc_post),
           legacy.axes = TRUE, size = 2, linetype = 2) +
  scale_x_continuous(labels = scales::percent) +
  scale_y_continuous(labels = scales::percent) +
  xlab("100% - specificity : False Positive Percentage") +
  ylab("Sensitivity : True Postive Percentage") + 
  theme_bw() +
  ggtitle(paste0("PRE AUC:", round(roc_pre$auc, 2), " POST AUC : ", round(roc_post$auc, 2))) +
  guides(color = guide_legend(title="Treatment")) +
  theme_bw(base_size = 20) +
    theme(axis.text = element_text(size = 20),
          axis.title = element_text(size = 20, face="bold"),
          legend.text = element_text(size = rel(1)),
          panel.border = element_blank(),
          panel.grid.minor = element_blank(),
          panel.grid.major = element_blank())

saveRDS(fig3d, "data/fig3d.RDS")
fig3d

```

# CI and roc.test
```{r}
ci.auc(roc_pre)
ci.auc(roc_post)
roc.test(roc_post, roc_pre)
```

\newpage

# R session

```{r}
sessionInfo()
```

