---
title: "DE analysis"
author: "Sergey Naumenko"
date: "`r Sys.Date()`"
output:
  pdf_document:
    toc: yes
  html_document:
    code_folding: hide
    df_print: paged
    highlights: pygments
    number_sections: yes
    self_contained: yes
    theme: default
    toc: yes
    toc_float:
      collapsed: yes
      smooth_scroll: yes
---

# Overview
- **paired multifactor test version**
- Principal Investigator: Beth Overmoyer
- Experiment: RNAseq_analysis_of_inflammatory_breast_cancer_hbc04141
- study 6 was excluded because if low read depth in 3373-3
- https://www.bioconductor.org/packages/release/bioc/vignettes/DEGreport/inst/doc/DEGreport.html
- [AnnotationHub](https://bioconductor.org/packages/release/bioc/html/AnnotationHub.html). We use ensembl version matching bcbio pipeline - v94.
- [HBC materials](https://hbctraining.github.io/DGE_workshop_salmon/lessons/03_DGE_QC_analysis.html)
- [HBC materials - functional analysis](https://hbctraining.github.io/DGE_workshop_salmon/lessons/functional_analysis_2019.html)
- http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
- https://support.bioconductor.org/p/62357/#62368
- https://support.bioconductor.org/p/84241/
- https://support.bioconductor.org/p/98628/

```{r include = FALSE}
remove_cases_2_19 <- FALSE
rebuild_rds <- FALSE

## Setup
### Bioconductor and CRAN libraries used
library(DESeq2)
library(tidyverse)
library(RColorBrewer)
library(pheatmap)
library(DEGreport)
library(tximport)
library(ggplot2)
library(ggrepel)
library(knitr)
library(AnnotationHub)
library(ensembldb)
library(ggplotify)
library(writexl)
library(clusterProfiler)

ggplot2::theme_set(theme_light(base_size = 14))

opts_chunk[["set"]](
    cache = FALSE,
    dev = c("png", "pdf"),
    error = TRUE,
    highlight = TRUE,
    message = FALSE,
    prompt = FALSE,
    tidy = FALSE,
    warning = FALSE)
```
```{r include = FALSE}
# Have a folder called `data`, and copy your Salmon folders here from the cluster. 
## List all directories containing data  
### change the pattern to something specific to your Salmon folders
samples <- list.files(path = "./data/final", 
                      full.names = T,
                      pattern = "^S")

## Obtain a vector of all filenames including the path
files <- file.path(samples, "salmon", "quant.sf")
files

## Since all quant files have the same name it is useful to have names for each element
### change the string in str_replace so the pattern matches your filenames
names(files) <- str_replace(samples, "./data/final/", "") 
```

```{r include = FALSE}
# Load the data and metadata
meta <- read_csv("tables/metadata_corrected.csv") %>% 
      column_to_rownames(var = "samplename") 

meta$response <- str_replace(meta$response, "Yes", "pCR")
meta$response <- str_replace(meta$response, "No", "non-pCR")

protein_coding_genes <- read_csv("tables/ensembl_w_description.protein_coding.csv")
```

```{r include = FALSE}
annotation_file <- "data/ensembl94_hg38_annotations.txt"
gene_symbol_file <- "data/gene_symbol.txt"
if (file.exists(annotation_file)){
    hsdb <- read_tsv(annotation_file)
    gene_symbol <- read_tsv(gene_symbol_file)
}else{
    # Connect to AnnotationHub
    ah <- AnnotationHub()
    # Query AnnotationHub
    hs_ens <- query(ah, c("Homo sapiens", "EnsDb"))

    # Get Ensembl94 - used in bcbio
    hs_ens <- hs_ens[["AH64923"]]

    # Extract gene-level information
    txdb <- transcripts(hs_ens,
                    return.type = "data.frame") %>% 
    dplyr::select(tx_id, gene_id)  

    genedb <- genes(hs_ens, 
              return.type = "data.frame") %>% 
    dplyr::select(gene_id, gene_name, symbol)  

    gene_symbol <- genedb %>% dplyr::select(gene_id, symbol)
    write_tsv(gene_symbol, gene_symbol_file)

    hsdb <- inner_join(txdb, genedb)
    write.table(hsdb, 
            file = annotation_file , 
            sep = "\t", 
            row.names = F, 
            quote = F)
}
```

```{r include = FALSE}
# Read in  a tx2gene file with transcript identifiers in the first column and gene identifiers in the second column
tx2gene <- hsdb[, c("tx_id", "gene_id")]

# Run tximport
txi_file <- "data/txi.RDS"
if (!rebuild_rds & file.exists(txi_file)){
    txi <- readRDS(txi_file)
}else{
    txi <- tximport(files, 
                type = "salmon", 
                tx2gene = tx2gene, 
                countsFromAbundance = "lengthScaledTPM", 
                ignoreTxVersion = FALSE)
    saveRDS(txi, txi_file)
    library(R.utils)
    txi$abundance %>% 
          as.data.frame() %>% 
          rownames_to_column(var = "ensembl_gene_id") %>% 
          write_csv("tables/counts.tpm.csv")
    gzip("tables/counts.tpm.csv")
    txi$counts %>% 
          as.data.frame() %>% 
          rownames_to_column(var = "ensembl_gene_id") %>% 
          write_csv("tables/counts.raw.csv")
    gzip("tables/counts.raw.csv")
}

# Look at the counts
class(txi)
attributes(txi)
txi$counts %>% View()
```


```{r, include = FALSE}
# Checking to see that the transcript to gene mapping is correct
#When you have annotations that are from a different source from your reference you can run into #problems (i.e lose genes). Some checks you can do before proceeding:

#1. Look at the dimensions of your count matrix. Do you have ~20k genes present? `dim(txi$counts)`
#2. When running `tximport()` you will get a message in your console. If you see something like #`transcripts missing from tx2gene` start troubleshooting.
dim(txi$counts)
```


```{r, include = FALSE}
# Sanity check that metadata matches your expression
#It is always a good idea to check if:
#1. Do you have expression data for all samples listed in your metadata?
#2. Are the samples in your expression data in the same order as your metadata?
### Check that sample names match in both files
all(colnames(txi$counts) %in% rownames(meta))
### Check that sample names match in both files
all(colnames(txi$counts) %in% rownames(meta))
### Check that all samples are in the same order
meta <- meta[colnames(txi$counts),]
all(colnames(txi$counts) == rownames(meta))
```
```{r, include = FALSE}

# Run DESeq2

#estimating size factors
#estimating dispersions
#gene-wise dispersion estimates
#mean-dispersion relationship
#final dispersion estimates
#fitting model and testing


#* [Estimating size factors and count normalization](https://hbctraining.github.io/DGE_workshop_salmon/lessons/02_DGE_count_normalization.html)
#* [Gene-wise dispersions](https://hbctraining.github.io/DGE_workshop_salmon/lessons/04_DGE_DESeq2_analysis.html)
#* [Mean-dispersion(variance) relationship and the Negative Binomial Model](https://hbctraining.github.io/DGE_workshop_salmon/lessons/01_DGE_setup_and_overview.html#modeling-count-data)
#* [Model fitting and hypothesis testing](https://hbctraining.github.io/DGE_workshop_salmon/lessons/05_DGE_DESeq2_analysis2.html)
```


```{r include = FALSE}
# Wald test
# *Here we subset protein coding genes*.
## Create DESeq2Dataset object
dds_file <- "data/dds.paired.RDS"

meta$treatment <- as.factor(meta$treatment)
meta$treatment <- relevel(meta$treatment, "pre")

meta$response <- as.factor(meta$response)
meta$er <- as.factor(meta$er)
meta$date_of <- as.factor(meta$date_of)
meta$tumor_percentage <- as.factor(meta$tumor_percentage)
meta$tumor_percentage_high <- as.factor(meta$tumor_percentage_high)
meta$study_id <- as.factor(meta$study_id)

meta <- meta %>% dplyr::arrange(response, study_id)
# add a factor to control for study id within response
# it is 26 x 18 samples so further matrix correction needed
meta$study_id_nested <- factor(c(rep(1:13, each = 2), rep(1:9, each = 2)))

meta <- meta[colnames(txi$counts),]
all(colnames(txi$counts) == rownames(meta))

# not removing them anymore - NA -> No in responsce for 2, 19
if (remove_cases_2_19){
    non_responders <- meta %>% dplyr::filter(study_id %in% c(2, 19)) %>% row.names() 
}

if(!rebuild_rds & file.exists(dds_file)){
    dds <- readRDS(dds_file)
}else{
    dds <- DESeqDataSetFromTximport(txi,
                                    colData = meta, 
                                    design = ~treatment)
    
    if (remove_cases_2_19){
         dds <- dds[,!colnames(dds) %in% non_responders]
    }
    #design(dds) <- formula(~ response + er + tumor_percentage_high + date_of + response:study_id_nested + response:treatment)
    design(dds) <- formula(~response + response:study_id_nested + response:treatment)
    
    # remove zero columns
    mm <- model.matrix(design(dds), meta)
    idx <- which(colSums(mm == 0) == nrow(mm))
    mm <- mm[, -idx]
    
    # subset protein-coding genes
    pc_genes <- intersect(protein_coding_genes$ensembl_gene_id, row.names(dds))
    dds <- dds[pc_genes,]
    dds <- estimateSizeFactors(dds)
    nc <- counts(dds, normalized = TRUE)
    # more stringent filter needed in the paired test, otherwise not converging
    # https://support.bioconductor.org/p/65091/
    filter <- rowSums(nc >= 30) >= 30
    dds <- dds[filter,]
    
    # Run DESeq2
    dds <- DESeq(dds, full = mm, betaPrior = FALSE)
    saveRDS(dds, dds_file)
}
counts <- counts(dds, normalized = TRUE)
```



```{r, include = FALSE}
#\newpage
# Results
# Get results for treatment
resultsNames(dds)

# return mean counts for a group of sample in a column
get_counts_for_samples <- function(samples, column_name){
    tpm_counts <- txi$abundance %>%
        as.data.frame() %>% 
        dplyr::select(any_of(samples)) %>% 
        rowMeans() %>% 
        as.data.frame() %>% 
        rownames_to_column(var = "ensembl_gene_id")
    colnames(tpm_counts) <- c("ensembl_gene_id", "tpm")
    
    tpm_counts <- tpm_counts %>% 
        dplyr::mutate("{column_name}" := round(tpm, 2)) %>% 
        dplyr::select(-tpm)
    
    return(tpm_counts)
}

# get rid of excess precision
comb_de_result_table <- function(results){
    results <- results %>%
        mutate(baseMean = round(baseMean, 2),
               log2FoldChange = round(log2FoldChange, 2),
               lfcSE = round(lfcSE, 2),
               stat = round(stat, 2),
               pvalue = format(pvalue, scientific = TRUE, digits = 2),
               padj = format(padj, scientific = TRUE, digits = 2))
    return(results)
}

samples_npCR_pre <- meta %>% dplyr::filter(response == "non-pCR" &
                                            treatment == "pre") %>% 
                    row.names() 
tpm_npCR_pre <- get_counts_for_samples(samples_npCR_pre, "non-pCR_pre_mean_tpm" )


samples_npCR_post <- meta %>% dplyr::filter(response == "non-pCR" & 
                                             treatment == "post") %>% 
                     row.names() 
tpm_npCR_post <- get_counts_for_samples(samples_npCR_post, "non-pCR_post_mean_tpm" )


tpm_counts <- tpm_npCR_pre %>% 
              left_join(tpm_npCR_post,
                        by = c("ensembl_gene_id" = "ensembl_gene_id"))

samples_pCR_pre <- meta %>% dplyr::filter(response == "pCR" &
                                            treatment == "pre") %>% 
                    row.names() 
tpm_pCR_pre <- get_counts_for_samples(samples_pCR_pre, "pCR_pre_mean_tpm" )


samples_pCR_post <- meta %>% dplyr::filter(response == "pCR" & 
                                             treatment == "post") %>% 
                     row.names() 
tpm_pCR_post <- get_counts_for_samples(samples_pCR_post, "pCR_post_mean_tpm" )


tpm_counts <- tpm_counts %>% 
          left_join(tpm_pCR_pre,
                        by = c("ensembl_gene_id" = "ensembl_gene_id")) %>% 
          left_join(tpm_pCR_post,
                        by = c("ensembl_gene_id" = "ensembl_gene_id"))


# 1425
res_pCR <- results(dds, name = "responsepCR") %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "gene") %>% 
  as_tibble() %>% 
  left_join(gene_symbol, by = c("gene" = "gene_id")) %>% 
  dplyr::filter(padj < 0.05) %>% 
  left_join(tpm_counts, by = c("gene" = "ensembl_gene_id")) %>% 
  comb_de_result_table()
  

# 2320
res_pCR_post <- results(dds, name = "responsepCR.treatmentpost") %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "gene") %>% 
  as_tibble() %>% 
  left_join(gene_symbol, by = c("gene" = "gene_id")) %>% 
  left_join(tpm_counts, by = c("gene" = "ensembl_gene_id"))

# for excel we need strings, for volcano - numbers
res_pCR_post_xls <- res_pCR_post %>% 
    dplyr::filter(padj < 0.05) %>% 
    comb_de_result_table()
       
# 2424
res_nonpCR_post <- results(dds,
                           name = "responsenon.pCR.treatmentpost") %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "gene") %>% 
  as_tibble() %>% 
  left_join(gene_symbol, by = c("gene" = "gene_id")) %>% 
  left_join(tpm_counts, by = c("gene" = "ensembl_gene_id"))
  
res_nonpCR_post_xls <- res_nonpCR_post %>% 
    dplyr::filter(padj < 0.05) %>% 
    comb_de_result_table()
    
# 23
res_pCR_diff <- results(dds, contrast = list("responsepCR.treatmentpost", "responsenon.pCR.treatmentpost")) %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "gene") %>% 
  as_tibble() %>% 
  left_join(gene_symbol, by = c("gene" = "gene_id")) %>% 
  dplyr::filter(padj < 0.05) %>% 
  left_join(tpm_counts, by = c("gene" = "ensembl_gene_id")) %>% 
  mutate(baseMean = round(baseMean, 2),
         log2FoldChange = round(log2FoldChange, 2),
         lfcSE = round(lfcSE, 2),
         stat = round(stat, 2),
         pvalue = format(pvalue, scientific = TRUE, digits = 2),
         padj = format(padj, scientific = TRUE, digits = 2))


# Separate into up and down-regulated gene sets
sig_pCR_post_up <- res_pCR_post %>% 
                       dplyr::filter(padj < 0.01 & log2FoldChange > 0) %>% 
                       pull(gene)

sig_pCR_post_down <- res_pCR_post %>% 
                       dplyr::filter(padj < 0.01 & log2FoldChange < 0) %>% 
                       pull(gene)


sig_nonpCR_post_up <- res_nonpCR_post %>% 
                       dplyr::filter(padj < 0.01 & log2FoldChange > 0) %>% 
                       pull(gene)

sig_nonpCR_post_down <- res_nonpCR_post %>% 
                       dplyr::filter(padj < 0.01 & log2FoldChange < 0) %>% 
                       pull(gene)


```


```{r, fig.width = 10, fig.height = 10, include = FALSE}
# pCR - post
# Add a column for significant genes
res_pCR_post_vis <- res_pCR_post %>% mutate(threshold = padj < 0.01)

# don't show labels for lfc < 1.5
res_pCR_post_vis$symbol <- ifelse((abs(res_pCR_post_vis$log2FoldChange) > 1.5),   
                                res_pCR_post_vis$symbol, "") 


# don't show labels for non-significant genes
res_pCR_post_vis$symbol <- ifelse(res_pCR_post_vis$threshold,   
                                  res_pCR_post_vis$symbol, "")

# don't show labels for -2 < LFC < 0  - too many of them
res_pCR_post_vis$symbol <- ifelse(res_pCR_post_vis$log2FoldChange < 0 &
                                    res_pCR_post_vis$log2FoldChange > -2,   
                                  "",
                                  res_pCR_post_vis$symbol) 

# suggested by Jennifer
dont_show <- c("C20orf204", "B4GALNT4", "SMIM34B", "ARHGAP33", 
               "ATRNL1", "CROCC2", "CEL", "FLG", "KIAA1549L", 
               "AHRR", "SLC25A27", "CPLX1", "SNX10", "STX11",
               "FUCA1", "ACP5", "MS4A4A", "GM2A", "TYROBP", "GBP5", 
               "SASH3", "CTSL", "TFEC", "DNASE2", "IGSF6", "IFI30", "LGALS3",
               "FCER1G", "VNN2", "SELPLG", "GPX1", "TAGAP", "PRDX1", "SLAMF8",
               "SSR4", "LST1", "VMO1", "OMD", "PRB2", "TBX21", "LIPA", "COMMD3-BMI1",
               "SOWAHD", "LGALS2", "SLAMF6", "SH2D1A", "ALDH2", "PTPRCAP")

res_pCR_post_vis[res_pCR_post_vis$symbol %in% dont_show,]$symbol <- ""


dont_show_more <- c("C9orf152")
res_pCR_post_vis[res_pCR_post_vis$symbol %in% dont_show_more,]$symbol <- ""

fig5a <- ggplot(res_pCR_post_vis,
                aes(log2FoldChange, -log10(padj), label = symbol)) +
  geom_point(aes(colour = threshold)) +
  ggtitle("Response pCR post-treatment (day8)") +
  xlab("log2 fold change") + 
  ylab("-log10 adjusted p-value") +
  scale_x_continuous(limits = c(-10, 10)) +
  scale_y_continuous(limits = c(0, 7)) +
  theme(legend.position = "none",
        plot.title = element_text(size = rel(1.5), hjust = 0.5),
        axis.title = element_text(size = rel(1.25)),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank()) +
  geom_label_repel(aes(label = symbol), 
                  data = subset(res_pCR_post_vis, log2FoldChange > 0),
                      size = 5, 
                      min.segment.length = 0,
                      segment.size = 0.1,
                      segment.color = "grey50",
                      direction     = "both",
                      max.overlaps = Inf,
                      segment.linetype = 5,
                      max.iter =  Inf,
                      hjust = "right",
                      xlim = c(3, 10),
                      force = 5,
                      force_pull = 0
                  ) +
  geom_label_repel(aes(label = symbol), 
                  data = subset(res_pCR_post_vis, log2FoldChange < 0),
                      size = 5, 
                      min.segment.length = 0,
                      segment.size = 0.1,
                      segment.color = "grey50",
                      direction     = "both",
                      max.overlaps = Inf,
                      segment.linetype = 5,
                      max.iter =  Inf,
                      hjust = "left",
                      xlim = c(-9, -4),
                      force = 5,
                      force_pull = 0
                  )

 # nudge_x = 9 - subset(resResponse_tb_vis, log2FoldChange > 0)$log2FoldChange,
#force_pull = 0, # attraction between label and point
 #                     force = 0.5 # repulsion between labels
#point.padding = 1,,
#                  point.size = NA,
#                  arrow = arrow(length = unit(0.015, "npc")),
#                  segment.linetype = 1,
#                  box.padding = 0.5,
#                  force_pull = 0,
#                  force = 1

saveRDS(fig5a, "data/fig5a.RDS")  
```

\newpage
```{r, fig.width = 10, fig.height = 10}
fig5a
```





```{r, fig.width = 10, fig.height = 10, include = FALSE}
# non_pCR - post
# Add a column for significant genes
res_nonpCR_post_vis <- res_nonpCR_post %>% mutate(threshold = padj < 0.01)

# don't show labels for lfc < 1.5
res_nonpCR_post_vis$symbol <- ifelse((abs(res_nonpCR_post_vis$log2FoldChange) > 1.5),   
                                res_nonpCR_post_vis$symbol, "") 


# don't show labels for non-significant genes
res_nonpCR_post_vis$symbol <- ifelse(res_nonpCR_post_vis$threshold,   
                                  res_nonpCR_post_vis$symbol, "")

# suggested by Jennifer
dont_show <- c("C20orf204", "B4GALNT4", "SMIM34B", "ARHGAP33", 
               "ATRNL1", "CROCC2", "CEL", "FLG", "KIAA1549L", 
               "AHRR", "SLC25A27", "CPLX1", "SNX10", "STX11",
               "FUCA1", "ACP5", "MS4A4A", "GM2A", "TYROBP", "GBP5", 
               "SASH3", "CTSL", "TFEC", "DNASE2", "IGSF6", "IFI30", "LGALS3",
               "FCER1G", "VNN2", "SELPLG", "GPX1", "TAGAP", "PRDX1", "SLAMF8",
               "SSR4", "LST1", "VMO1", "OMD", "PRB2", "TBX21", "LIPA", "COMMD3-BMI1",
               "SOWAHD", "LGALS2", "SLAMF6", "SH2D1A", "ALDH2", "PTPRCAP")

res_nonpCR_post_vis[res_nonpCR_post_vis$symbol %in% dont_show,]$symbol <- ""


fig5b <- ggplot(res_nonpCR_post_vis,
       aes(log2FoldChange, -log10(padj), label = symbol)) +
  geom_point(aes(colour = threshold)) +
  ggtitle("Response non-pCR post treatment (day8)") +
  xlab("log2 fold change") + 
  ylab("-log10 adjusted p-value") +
  scale_x_continuous(limits = c(-10, 10)) +
  scale_y_continuous(limits = c(0, 7)) +
  theme(legend.position = "none",
        plot.title = element_text(size = rel(1.5), hjust = 0.5),
        axis.title = element_text(size = rel(1.25)),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank()) +
  geom_label_repel(aes(label = symbol), 
                  data = subset(res_nonpCR_post_vis, 
                                log2FoldChange > 0),
                      size = 5, 
                      min.segment.length = 0,
                      segment.size = 0.1,
                      segment.color = "grey50",
                      direction     = "both",
                      max.overlaps = Inf,
                      segment.linetype = 5,
                      max.iter =  Inf,
                      hjust = "right",
                      xlim = c(3, 10),
                      force = 5,
                      force_pull = 0
                  ) +
  geom_label_repel(aes(label = symbol), 
                  data = subset(res_nonpCR_post_vis, log2FoldChange < 0),
                      size = 5, 
                      min.segment.length = 0,
                      segment.size = 0.1,
                      segment.color = "grey50",
                      direction     = "both",
                      max.overlaps = Inf,
                      segment.linetype = 5,
                      max.iter =  Inf,
                      hjust = "left",
                      xlim = c(-9, -4),
                      force = 5,
                      force_pull = 0
                  )

 # nudge_x = 9 - subset(resResponse_tb_vis, log2FoldChange > 0)$log2FoldChange,
#force_pull = 0, # attraction between label and point
 #                     force = 0.5 # repulsion between labels
#point.padding = 1,,
#                  point.size = NA,
#                  arrow = arrow(length = unit(0.015, "npc")),
#                  segment.linetype = 1,
#                  box.padding = 0.5,
#                  force_pull = 0,
#                  force = 1

saveRDS(fig5b, "data/fig5b.RDS")
```

\newpage
```{r, fig.width = 10, fig.height = 10}
fig5b
```


```{r, fig.width = 20, fig.height = 30, echo = FALSE}
# Heatmaps
# Create a matrix of normalized expression

sig_up <- res_pCR_post_xls %>% arrange(-log2FoldChange) %>% head(50) %>% pull(gene)
sig_down <- res_pCR_post_xls %>% arrange(log2FoldChange) %>% head(50) %>% pull(gene)
sig <- c(sig_up, sig_down)

row_annotation <- gene_symbol %>% 
                    as_tibble() %>% 
                    dplyr::filter(gene_id %in% sig)

plotmat <- txi$abundance[c(sig_up, sig_down),] %>% 
    as.data.frame() %>% 
    rownames_to_column(var = "ensembl_gene_id") %>% 
    left_join(gene_symbol, by = c("ensembl_gene_id" = "gene_id")) %>% 
    drop_na(symbol)

plotmat$ensembl_gene_id <- NULL

plotmat <- plotmat %>% column_to_rownames(var = "symbol") %>% as.matrix()

# Color palette
heat.colors <- brewer.pal(6, "YlOrRd")

# Plot heatmap
# color = heat.colors,
pheatmap(plotmat, scale = "row", 
         show_rownames = TRUE,
         border = FALSE,
         annotation = meta[, c("treatment"), drop = FALSE],
         main = "Top 50 Up- and Down- regulated genes in pCR post (day8)",
         fontsize = 20)
```

```{r, fig.width = 20, fig.height = 30, echo = FALSE}
# Heatmaps
# Create a matrix of normalized expression

sig_up <- res_nonpCR_post_xls %>% arrange(-log2FoldChange) %>% head(50) %>% pull(gene)
sig_down <- res_nonpCR_post_xls %>% arrange(log2FoldChange) %>% head(50) %>% pull(gene)
sig <- c(sig_up, sig_down)

row_annotation <- gene_symbol %>% 
                    as_tibble() %>% 
                    dplyr::filter(gene_id %in% sig)

plotmat <- txi$abundance[c(sig_up, sig_down),] %>% 
    as.data.frame() %>% 
    rownames_to_column(var = "ensembl_gene_id") %>% 
    left_join(gene_symbol, by = c("ensembl_gene_id" = "gene_id")) %>% 
    drop_na(symbol)

plotmat$ensembl_gene_id <- NULL

plotmat <- plotmat %>% column_to_rownames(var = "symbol") %>% as.matrix()

# Color palette
heat.colors <- brewer.pal(6, "YlOrRd")

# Plot heatmap
# color = heat.colors,
pheatmap(plotmat, scale = "row", 
         show_rownames = TRUE,
         border = FALSE,
         annotation = meta[, c("treatment"), drop = FALSE],
         main = "Top 50 Up- and Down- regulated genes in non-pCR post (day8)",
         fontsize = 20)
```



```{r, fig.width = 12, fig.height = 12, echo = FALSE}
# Functional analysis
## Biological Process (BP)
bg_genes <- res_pCR_post$gene

## Run GO enrichment analysis 
compgo_file <- "data/pCR_post.compgo.up.RDS"
if (file.exists(compgo_file)){
    compGO <- readRDS(compgo_file)
}else{
    compGO <- enrichGO(gene = sig_pCR_post_up,
                   universe = bg_genes,
                   keyType = "ENSEMBL",
                   OrgDb = "org.Hs.eg.db", 
                   ont = "BP", 
                   qvalueCutoff  = 0.05, 
                   pAdjustMethod = "BH",
                   readable = TRUE)
    saveRDS(compGO, compgo_file)
}

pcr_post_go_bp_up <- data.frame(compGO@result) %>% dplyr::filter(p.adjust < 0.05)

dotplot(compGO,
        showCategory = 20, 
        title = "GO (Biological Process) Enrichment Analysis for UP in pCR - post (day8)",
        label_format = 20,
        font.size = 10)
```


\newpage

```{r, fig.width = 12, fig.height = 12, echo = FALSE}
compgo_file <- "data/pCR_post.compgo.down.RDS"
if (file.exists(compgo_file)){
    compGO <- readRDS(compgo_file)
}else{
    compGO <- enrichGO(gene = sig_pCR_post_down,
                   universe = bg_genes,
                   keyType = "ENSEMBL",
                   OrgDb = "org.Hs.eg.db", 
                   ont = "BP", 
                   qvalueCutoff  = 0.05, 
                   pAdjustMethod = "BH",
                   readable = TRUE)
    saveRDS(compGO, compgo_file)
}

pcr_post_go_bp_down <- data.frame(compGO@result) %>% dplyr::filter(p.adjust < 0.05)
#print("Down")
#nrow(results_down)
#write_csv(results_down, "tables/paired.pCR_post.GO_BP_DOWN.csv")
dotplot(compGO,
        showCategory = 20, 
        title = "GO (Biological Process) Enrichment Analysis for Down in pCR - post (Day8)",
        label_format = 20,
        font.size = 10)
```

\newpage

```{r, fig.width = 12, fig.height = 12, echo = FALSE}
# Functional analysis
## Biological Process (BP)
bg_genes <- res_nonpCR_post$gene

## Run GO enrichment analysis 
compgo_file <- "data/nonpCR_post.compgo.up.RDS"
if (file.exists(compgo_file)){
    compGO <- readRDS(compgo_file)
}else{
    compGO <- enrichGO(gene = sig_nonpCR_post_up,
                   universe = bg_genes,
                   keyType = "ENSEMBL",
                   OrgDb = "org.Hs.eg.db", 
                   ont = "BP", 
                   qvalueCutoff  = 0.05, 
                   pAdjustMethod = "BH",
                   readable = TRUE)
    saveRDS(compGO, compgo_file)
}

nonpcr_post_go_bp_up <- data.frame(compGO@result) %>% dplyr::filter(p.adjust < 0.05)

dotplot(compGO,
        showCategory = 20, 
        title = "GO (Biological Process) Enrichment Analysis for UP in non-pCR - post (day8))",
        label_format = 20,
        font.size = 10)
```


\newpage

```{r, fig.width = 12, fig.height = 12, echo = FALSE}
compgo_file <- "data/nonpCR_post.compgo.down.RDS"
if (file.exists(compgo_file)){
    compGO <- readRDS(compgo_file)
}else{
    compGO <- enrichGO(gene = sig_nonpCR_post_down,
                   universe = bg_genes,
                   keyType = "ENSEMBL",
                   OrgDb = "org.Hs.eg.db", 
                   ont = "BP", 
                   qvalueCutoff  = 0.05, 
                   pAdjustMethod = "BH",
                   readable = TRUE)
    saveRDS(compGO, compgo_file)
}

nonpcr_post_go_bp_down <- data.frame(compGO@result) %>% dplyr::filter(p.adjust < 0.05)
#print("Down")
#nrow(results_down)
#write_csv(results_down, "tables/paired.pCR_post.GO_BP_DOWN.csv")
dotplot(compGO,
        showCategory = 20, 
        title = "GO (Biological Process) Enrichment Analysis for Down in non-pCR post (day8))",
        label_format = 20,
        font.size = 10)
```



```{r, include = FALSE}

non_paired_T13 <- read_csv("tables/T13.DE_response_day8.csv")
intersect_non_paired_T13 <- res_pCR_post_xls %>% 
    semi_join(non_paired_T13, by = c("gene" = "gene"))

non_paired_T4 <- read_csv("tables/T4.DE_response.csv")
intersect_non_paired_T4 <- res_pCR_post_xls %>% 
    semi_join(non_paired_T4, by = c("gene" = "gene"))

write_xlsx(list(T1.res_pCR = res_pCR,
                T2.res_pCR_post = res_pCR_post_xls, 
                T3.res_nonpCR_post = res_nonpCR_post_xls, 
                T4.res_pCR_diff = res_pCR_diff,
                T5.pCR_post_GO_BP_UP = pcr_post_go_bp_up,
                T6.pCR_post_GO_bp_down = pcr_post_go_bp_down,
                T7.non_pCR_post_GO_BP_UP = nonpcr_post_go_bp_up,
                T8.non_pCR_post_GO_bp_down = nonpcr_post_go_bp_down,
                T9.res_pCR_post_intersect_T13_non_paired = intersect_non_paired_T13,
                T10.res_pCR_post_intersect_T4_non_paired = intersect_non_paired_T4), 
           path = "tables/paired_test.xlsx")

```


\newpage
# R session
```{r}
sessionInfo()
```
